docs/Blog-Post.md

```diff
@@ -0,0 +1,525 @@
+and offers strong reproducibility guarantees. JAX also works well on TPUs, while we found that PyTorch support was still uneven.
+JAX is also a natural choice because it allows you to focus on the "what" of your code, and not on the "how": details of
+partitioning and communication can be left to the XLA compiler. Finally, JAX makes reproducibility easy, since it uses
+bitwise deterministic PRNGs by default, with careful control over the PRNG state.
+
+However, JAX is a low-level framework, and we found that, by itself, it did not provide the legibility that we wanted.
+
+For us, legibility is a top concern. We therefore created two new libraries: **Haliax** and **Levanter**. Haliax is a
+named tensor library that focuses on improving the legibility and compositionality of deep learning code while still
+maintaining efficiency and scalability. Levanter is a library for training foundation models built on top of Haliax. In
+addition to the goals of legibility, efficiency, and scalability, Levanter further strives for bitwise reproducibility,
+meaning that the same code with the same data will produce the exact same result, even in the presence of preemption and
+restarts.
+
+# Haliax: Legibility via Named Tensors
+
+Haliax is a library for named tensors, built on JAX and [Equinox](https://github.com/patrick-kidger/equinox),
+which is a neural network library for JAX that provides a familiar, PyTorch-like module structure. Haliax uses
+Equinox's module structure for its neural network library, rather than Flax or Haiku.
+
+Named Tensors are a powerful abstraction that allow you to give names to the axes of your tensors. These names help
+make your code more legible, more composable, and less bug-prone. In Haliax, they also form the basis of how we handle
+scale with [Fully-Sharded Data Parallel](https://engineering.fb.com/2021/07/15/open-source/fsdp/) and tensor parallelism.
+(See below for our tutorial!)
+
+Haliax is modeled on Alexander Rush's [Tensor Considered Harmful](https://nlp.seas.harvard.edu/NamedTensor),
+which argues that named tensors are a better abstraction than the positional axes that are common in deep learning.
+In particular, he argues that:
+
+* Named axes are more semantically meaningful and rely less on bitrot-prone comments.
+* Named axes allow you to abstract over unreferenced dimensions, making code more flexible.
+* Broadcasting leads to unreadable strings of `view`s and `squeeze`s.
+
+To this, I'd add that the implicit broadcasting so common in deep learning code is a source of easy-to-miss bugs, and
+that named tensors eliminate many of these bugs.
+
+## A Quick Example: Attention in Haliax
+
+This blog post isn't the place for a full introduction to Haliax (please see the [Haliax tutorial](https://colab.research.google.com/drive/1TiTcQQ4V5mopbgCu1SVl-oqJtXn7rFnC)),
+but here's a quick example of a minimal, but full-featured, attention implementation in Haliax.
+
+```python
```

Comment:
Can we do a diff between what this would look like without named tensors?

Reply:
how about I literally quote the mingpt implementation, which is an implementation that claims to care about legibility, but looks deranged.

Reply:
(did this, line 238 ff)